content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
++++
+title = "DDoS Attack Detection with Wallaroo: A Real-time Time Series Analysis Example"
+draft = false
+date = 2017-11-24T00:00:00Z
+tags = [
+    "wallaroo",
+    "windowing",
+    "partitioning",
+    "python",
+    "example",
+    "tutorial",
+    "ddos",
+    "timeseries",
+    "anomaly detection"
+]
+categories = [
+    "Windowing",
+    "Examples"
+]
+description = "Building a DDoS attack detector using timeseries analysis in real time with Wallaroo"
+author = "nisanharamati"
++++
+
+
+This post will go through a real-world use case for Wallaroo, our distributed data processing framework for building high-performance streaming data applications. We will construct a streaming DDoS attack detector, which consumes a stream of request logs from a large group of servers and uses statistical anomaly detection to alert us when a server is under attack.
+
+If you're new to Wallaroo and are interested in learning more about it before diving into today's example, you can check our [Go Python, Go! Stream Processing for Python][go python] post, where we introduce the Wallaroo Python API.
+
+The code for the example in this post can be found on [GitHub][ddos-detection], and can be run with the [current version of Wallaroo (0.2.2)](https://github.com/WallarooLabs/wallaroo/tree/0.2.2).
+
+## When Your Service Layer is Under Attack
+
+Running a large application over the internet is really great! You can reach a massive customer base, and provide them with your digital product in no time at all. Unless...
+
+![this site cannot be reached][this site cannot be reached]
+
+According to [Cloudflare][cloudflare what is ddos],
+
+> [a] distributed denial-of-service (DDoS) attack is a malicious attempt to disrupt normal traffic of a targeted server, service or network by overwhelming the target or its surrounding infrastructure with a flood of Internet traffic.
+
+The attack utilizes the fact that responding to each request takes up a fraction of your system's finite resources. So if an attacker can generate sufficiently many requests, they can use up all of your resources, leaving none available to service legitimate requests from your users. And in order to avoid being throttled, an attacker will use a large number of (probably) compromised machines to send service requests, making it hard to block them.
+
+Responding to attacks is an art in itself, but first you must know that it's happening. Early detection is key.
+
+## Detecting that an Attack is Underway
+
+In order to detect an attack, we need two things: a model for "good" or "normal" behaviour, and a way to measure our services' current behaviour to compare against the "normal" one. If the difference is too great, we can mark the present data point as anomalous, and in the DDoS attack case, we will mark the server as "under attack". This process is called anomaly detection.
```

Comment:
I think the "two things we need" should be a list. It will make it easier to quickly absorb the info.

Reply:
Also, "the difference" might not be obvious. I think saying "the difference between normal and current" would work.

" If the difference between normal and current is too great, we can mark the present data point as anomalous. If anomalous behavior is observed, it can be analyzed in attempt to determine the cause. A DDoS attack is one possible cause. If we determine we are under attack, we will mark the server as "under attack".